22-perf/post.md

```diff
@@ -0,0 +1,581 @@
+You don't trust me? You don't need to. Let's modify the program to use a channel with a buffer
+of a size that allows us to fit all of the pixels.
+
+[embedmd]:# (main.go /func createWorkersBuffered/ /make\(chan.*/)
+```go
+func createWorkersBuffered(width, height int) image.Image {
+	m := image.NewGray(image.Rect(0, 0, width, height))
+
+	type px struct{ x, y int }
+	c := make(chan px, width*height)
+```
+
+The rest of the function is exactly as the previous one.
+
+Will this make our program faster? Let's see!
+
+```bash
+$ time ./mandelbrot
+./mandelbrot  8.55s user 0.37s system 494% cpu 1.804 total
+```
+
+It is much faster than before but still not nearly as fast as creating one goroutine per row.
+
+The tracer shows that we still have some contention, and pprof quantifies it to 613ms.
+
+```bash
+$ time ./mandelbrot > workers-buf.trace
+./mandelbrot > workers-buf.trace  8.06s user 0.29s system 542% cpu 1.539 total
+
+$ go tool trace workers-buf.trace
+2017/10/23 12:04:51 Parsing trace...
+2017/10/23 12:04:51 Serializing trace...
+2017/10/23 12:04:51 Splitting trace...
+2017/10/23 12:04:51 Opening browser
+```
+
+![workers buffered sync](img/workers-buf-sync.png)
+
+## Compromises
+
+Sending values through channels has a very low cost, but it does have a cost and we're doing this
+over 4 million times in this program.
+
+It seems like a good idea to limit this cost by creating fewer tasks. How can we even do that?
+Each task should be larger: let's create a task per row.
+
+[embedmd]:# (main.go /func createRowWorkers/ /^}/)
+```go
+func createRowWorkers(width, height int) image.Image {
+	m := image.NewGray(image.Rect(0, 0, width, height))
+
+	c := make(chan int)
+
+	var w sync.WaitGroup
+	w.Add(8)
+	for i := 0; i < 8; i++ {
+		go func() {
+			for i := range c {
+				for j := 0; j < height; j++ {
+					m.Set(i, j, pixel(i, j, width, height))
+				}
+			}
+			w.Done()
+		}()
+	}
+
+	for i := 0; i < width; i++ {
+		c <- i
+	}
+
+	close(c)
+	w.Wait()
+	return m
+}
+```
+
+How much faster is this?
+
+```bash
+$ time ./mandelbrot
+./mandelbrot  4.72s user 0.04s system 495% cpu 0.962 total
+```
+
+Pretty fast! And if we added a buffer to the channel, as shown in `createRowWorkersBuffered`?
+
+```bash
+$ time ./mandelbrot
+./mandelbrot  4.74s user 0.04s system 513% cpu 0.931 total
+```
+
+Just a tiny bit better.
+
+# Conclusion
+
+So, it seems like creating one goroutine per row is the fastest option, but probably this changes
+given different image sizes. It is time to write benchmarks (they're already written in the repo!),
+look at the traces and pprof data and make sensible and informed decisions.
+
+We saw how pprof allows us to see how our program uses the CPU and how much locking contention exists.
+Pprof also provides a way to profile memory allocations which we didn't cover.
+On the other side we saw the tracer, that allows us to have a very detailed view of what our program
+does, specially when our program doesn't do much!
```

Comment:
It would be nice to add links to articles about the tools : 
- Pprof: https://blog.golang.org/profiling-go-programs  (this is from 2011, a bit old)
- Trace: https://making.pusher.com/go-tool-trace/  (I find it very good)